src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -304,14 +306,18 @@ impl<O: ForestObligation> ObligationForest<O> {
     }
 
     // Returns Err(()) if we already know this obligation failed.
-    fn register_obligation_at(&mut self, obligation: O, parent: Option<usize>) -> Result<(), ()> {
-        if self.done_cache.contains(obligation.as_predicate()) {
-            return Ok(());
-        }
-
+    fn register_obligation_at(
+        &mut self,
+        obligation: O,
+        parent: Option<NodeIndex>,
+    ) -> Result<(), ()> {
         match self.active_cache.entry(obligation.as_predicate().clone()) {
```

Comment:
This now clones the predicate even when it is done. Maybe that is the problem?

Reply:
Looked into it the predicate type, it is always a `ty::Predicate` (or `&str` in tests) which is copyable and only takes up 32 bytes which isn't tiny but shouldn't really affect things...

Reply:
Ugh, ok I think it is because `active_cache` now gets really big, so the `retain` becomes really expensive as it fills up with `None` values...